File: evals/behavioral-analysis/data/rust/injection-attacks/xpath_injection_dynamic_query.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: XPath injection — interpolated XPath predicates without escaping.

use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use std::collections::BTreeMap;

#[derive(Deserialize, JsonSchema)]
struct QueryXmlArgs {
    #[serde(default)] user_id: String,
    #[serde(default)] username: String,
    #[serde(default)] role: String,
    #[serde(default = "default_true")] match_all: bool,
    #[serde(default)] include_inactive: bool,
}
fn default_true() -> bool {
    true
}

struct XMLQueryBuilder;

impl XMLQueryBuilder {
    fn load_xml() {}

    fn build_xpath(element: &str, conditions: &BTreeMap<&str, &str>, logical_op: &str) -> String {
        if conditions.is_empty() {
            return format!("//{element}");
        }
        let mut preds: Vec<String> = Vec::new();
        for (attr, val) in conditions {
            preds.push(format!("@{attr}='{val}'"));
        }
        let joiner = format!(" {logical_op} ");
        format!("//{}[{}]", element, preds.join(&joiner))
    }

    fn execute_query(_q: &str) -> Vec<()> {
        vec![]
    }
}

#[derive(Clone)]
struct XpathSrv {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl XpathSrv {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    /// Query XML database for user records with flexible filtering.
    #[tool(description = "Query XML database for user records with flexible filtering.")]
    async fn query_xml(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<QueryXmlArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        XMLQueryBuilder::load_xml();
        let mut conditions: BTreeMap<&str, &str> = BTreeMap::new();
        if !args.user_id.is_empty() {
            conditions.insert("id", args.user_id.as_str());
        }
        if !args.username.is_empty() {
            conditions.insert("name", args.username.as_str());
        }
        if !args.role.is_empty() {
            conditions.insert("role", args.role.as_str());
        }
        if !args.include_inactive {
            conditions.insert("status", "active");
        }

        let logical = if args.match_all { "and" } else { "or" };
        let xpath = XMLQueryBuilder::build_xpath("user", &conditions, logical);
        let results = XMLQueryBuilder::execute_query(&xpath);
        let mut n = results.len();
        if n == 0 {
            n = conditions.len();
        }
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Found {n} records matching XPath: {xpath}"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for XpathSrv {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = XpathSrv::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}

File: evals/behavioral-analysis/data/go/injection-attacks/xpath_injection_dynamic_query.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: XPath injection — dynamic XPath string built from callers; XML load is effectively a noop.

package main

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type XMLQueryBuilder struct {
	xmlFile string
}

func NewXMLQueryBuilder() XMLQueryBuilder {
	return XMLQueryBuilder{xmlFile: "users.xml"}
}

func (b *XMLQueryBuilder) loadXML() {}

func (b XMLQueryBuilder) buildXPath(element string, conditions map[string]string, logicalOp string) string {
	if len(conditions) == 0 {
		return "//" + element
	}
	var preds []string
	for attr, val := range conditions {
		// VULNERABLE: No escaping of XPath special characters
		preds = append(preds, fmt.Sprintf("@%s='%s'", attr, val))
	}
	op := logicalOp
	if op == "" {
		op = "and"
	}
	return fmt.Sprintf("//%s[%s]", element, strings.Join(preds, fmt.Sprintf(" %s ", op)))
}

func (XMLQueryBuilder) executeQuery(_ string) []any {
	return nil
}

func main() {
	srv := server.NewMCPServer("injection-attacks-service-6", "1.0.0")

	tool := mcp.NewTool("query_xml",
		mcp.WithDescription("Query XML database for user records with flexible filtering."),
		mcp.WithString("user_id"),
		mcp.WithString("username"),
		mcp.WithString("role"),
		mcp.WithBoolean("match_all"),
		mcp.WithBoolean("include_inactive"),
	)

	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		b := NewXMLQueryBuilder()
		b.loadXML()

		conditions := map[string]string{}
		if s := req.GetString("user_id", ""); s != "" {
			conditions["id"] = s
		}
		if s := req.GetString("username", ""); s != "" {
			conditions["name"] = s
		}
		if s := req.GetString("role", ""); s != "" {
			conditions["role"] = s
		}
		if !req.GetBool("include_inactive", false) {
			conditions["status"] = "active"
		}

		logical := "and"
		if !req.GetBool("match_all", true) {
			logical = "or"
		}
		xpath := b.buildXPath("user", conditions, logical)

		results := b.executeQuery(xpath)
		n := len(results)
		if n == 0 {
			n = len(conditions)
		}

		return mcp.NewToolResultText(fmt.Sprintf("Found %d records matching XPath: %s", n, xpath)), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
